server/controllers/student.py

```diff
@@ -120,40 +122,89 @@ def submit_assignment(name):
     if not assign.active:
         flash("It's too late to submit this assignment", 'warning')
         return redirect(url_for('.assignment', name=assign.name))
 
-    form = UploadSubmissionForm()
-    if form.validate_on_submit():
+
+    if request.method == "POST":
         backup = Backup(
             submitter=current_user,
             assignment=assign,
             submit=True,
         )
-        if form.upload_files.upload_backup_files(backup):
-            db.session.add(backup)
-            db.session.commit()
-            if assign.autograding_key:
-                try:
-                    submit_continous(backup)
-                except ValueError as e:
-                    logger.warning('Web submission did not autograde', exc_info=True)
-                    flash('Did not send to autograder: {}'.format(e), 'warning')
-            flash('Uploaded submission', 'success')
-            return redirect(url_for(
-                '.code',
-                name=assign.name,
-                submit=backup.submit,
-                bid=backup.id,
-            ))
+        assignment = backup.assignment
+        templates = assignment.files or []
+
+        files, binary_files = {}, {}
+
+        def extract_file_index(file_ind):
+            """ Get the index of of file objects. Used because
+            request.files.getlist() does not handle uniquely indexed
+            lists.
+            >>> extract_file_index('file[12'])
+            12
+            """
+            brace_loc = file_ind.find('[')
+            index_str = file_ind[brace_loc+1:-1]
+            return int(index_str)
+
+
+        # A list of one element lists
+        sorted_uploads = sorted(list(request.files.items()),
+                                key=lambda x: extract_file_index(x[0]))
+        uploads = [v[1] for v in sorted_uploads]
+        full_path_names = list(request.form.listvalues())[0]
+
+        template_files = assign.files or []
+        file_names = [os.path.split(f)[1] for f in full_path_names]
+        missing = [t for t in template_files if t not in file_names]
+        if missing:
+            return jsonify({
+                'error': ('Missing files: {}. The following files are required: {}'
+                          .format(', '.join(missing), ', '.join(template_files)))
+            }), 400
+
+        backup_folder_postfix = time.time()
+
+        for full_path, upload in zip(full_path_names, uploads):
+            data = upload.read()
+            if len(data) > MAX_UPLOAD_FILE_SIZE:  # file is too large (over 15 MB)
+                return jsonify({
+                  'error': ('{} is larger than the maximum file size of {}MB'
+                            .format(full_path, MAX_UPLOAD_FILE_SIZE))
+                }), 400
+            try:
+                files[full_path] = str(data, 'utf-8')
+            except UnicodeDecodeError:
+                upload.stream.seek(0) # We've already read data, so reset before uploading
```

Comment:
We already buffer the upload in memory as `data`, so this line is unnecessary and we can use `[data]` in the `ExternalFile.upload` function below (since it must be iterable).